File: day_1/second_star/haskell/app/Main.hs
import Data.Char (isDigit, digitToInt)
import Data.Maybe (fromMaybe)
import Data.List (find, isPrefixOf)

main :: IO ()
main = print . sum . map getNumbers . lines =<< readFile "../../input.txt"

getNumbers :: String -> Int
getNumbers str = read $ show (getDigit str) ++ show (getDigit $ reverse str) :: Int
  where 
    getDigit = fromMaybe 0 . getFirstDigit

getFirstDigit :: String -> Maybe Int
getFirstDigit [] = Nothing
getFirstDigit str@(x:xs)
  | isDigit x = Just $ digitToInt x
  | otherwise = case find ((`isPrefixOf` str) . fst) digits of
      Just (_, value) -> Just value
      Nothing -> getFirstDigit xs
  where
    digits = [("one", 1), ("two", 2), ("three", 3), ("four", 4), ("five", 5), 
              ("six", 6), ("seven", 7), ("eight", 8), ("nine", 9),
              ("eno", 1), ("owt", 2), ("eerht", 3), ("ruof", 4), ("evif", 5), 
              ("xis", 6), ("neves", 7), ("thgie", 8), ("enin", 9)]



-- import Data.Char (isDigit, digitToInt)
-- import Data.Maybe (fromMaybe)
-- import Data.List (find, isPrefixOf)

-- main :: IO ()
-- main = print . sum . map getNumbers . lines =<< readFile "../../input.txt"

-- getNumbers :: String -> Int
-- getNumbers str = read $ show (getDigit str) ++ show (getDigit $ reverse str) :: Int
--   where 
--     getDigit = fromMaybe 0 . getFirstDigit

-- getFirstDigit :: String -> Maybe Int
-- getFirstDigit [] = Nothing
-- getFirstDigit str@(x:xs)
--   | isDigit x = Just $ digitToInt x
--   | otherwise = case find ((`isPrefixOf` str) . fst) digits of
--       Just (_, value) -> Just value
--       Nothing -> getFirstDigit xs
--   where
--     digits = [("one", 1), ("two", 2), ("three", 3), ("four", 4), ("five", 5), 
--               ("six", 6), ("seven", 7), ("eight", 8), ("nine", 9),
--               ("eno", 1), ("owt", 2), ("eerht", 3), ("ruof", 4), ("evif", 5), 
--               ("xis", 6), ("neves", 7), ("thgie", 8), ("enin", 9)]

File: day_1/second_star/zig/src/main.zig
const std = @import("std");

pub fn main() !void {
    const input = try std.fs.cwd().openFile("../../input.txt", .{});
    defer input.close();

    var buffer: [4096]u8 = undefined;
    var reader = input.reader();

    const nums = [_][]const u8{
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    };

    var output: usize = 0;
    while (try reader.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
        var first: u32 = 0;
        var last: u32 = 0;
        var i: usize = 0;
        while (i < line.len) {
            var cur: ?u32 = null;
            // std.debug.print("{c}\n", .{line[i]});
            if (std.ascii.isDigit(line[i])) {
                cur = (line[i] - '0') % 10;
            } else {
                var j: usize = 0;
                while (j < nums.len) : (j += 1) {
                    if (std.mem.startsWith(u8, line[i..], nums[j])) {
                        cur = @as(u32, @truncate(j)) + 1;
                        break;
                    }
                }
            }
            if (cur) |c| {
                if (first == 0) {
                    first = c;
                    last = c;
                } else {
                    last = c;
                }
            }
            i += 1;
        }
        output += first * 10 + last;
    }

    std.debug.print("{}\n", .{output});
}
